optuna/visualization/_timeline.py

```diff
@@ -0,0 +1,134 @@
+import datetime
+from typing import List
+from typing import NamedTuple
+
+from optuna._experimental import experimental_func
+from optuna.logging import get_logger
+from optuna.study import Study
+from optuna.trial import FrozenTrial
+from optuna.trial import TrialState
+from optuna.visualization._pareto_front import _make_hovertext
+from optuna.visualization._plotly_imports import _imports
+
+
+if _imports.is_successful():
+    from optuna.visualization._plotly_imports import go  # px, go
+
+_logger = get_logger(__name__)
+
+
+class _TimelineBarInfo(NamedTuple):
+    number: int
+    start: datetime.datetime
+    end: datetime.datetime
+    state: TrialState
+    frozen_trial: FrozenTrial
```

Comment:
his class, only this `frozen_trial` attribute seems too granular. For example, why not read `_make_hovertext` first before creating this class and have that text as an attribute of str?

Reply:
Agreed. I revised the code in https://github.com/optuna/optuna/pull/4470/commits/beecdca130691b50ee2bd1a2deeb38432005f98e .